lib/yang.c

```diff
@@ -545,6 +545,47 @@ void yang_dnode_free(struct lyd_node *dnode)
 	lyd_free_withsiblings(dnode);
 }
 
+void yang_dnode_replace(struct lyd_node **dnode_dst, struct lyd_node *dnode_src)
+{
+	struct lyd_difflist *diff;
+
+	diff = lyd_diff(*dnode_dst, dnode_src, 0);
```

Comment:
I guess we talked about this during initial discussion about comparing nodes. How lighter is function lyd_diff v/s blindly freeing all the nodes and allocating new one?. The reason is we may need to look at how the candidate table is maintained today. If we keep only delta then we don't need to compare 1000 and 1001 entries.

Reply:
> I guess we talked about this during initial discussion about comparing nodes. How lighter is function lyd_diff v/s blindly freeing all the nodes and allocating new one?.

The difference is quite substantial. I lost the flamegraphs I generated when I did this work, but I'll try to generate new ones later and post them here.

> The reason is we may need to look at how the candidate table is maintained today. If we keep only delta then we don't need to compare 1000 and 1001 entries.

Maintaining only the deltas isn't possible using libyang today, since it needs full data trees to perform validation. But in practice this isn't a problem since `lyd_diff()` is really fast (due to their cache mechanism). Also, if you maintained only the deltas, you'd need to keep a copy of the running config anyway to prevent inconsistencies (other NB clients can still commit changes behind your back).

Reply:
I think this is a optimisation topic we may need to revisit and this PR changes here looks good. Running config should be maintained irrespective of candidate table is having full context or only delta. If candidate has one change and 1000 already in running config then we end up doing a diff on all 1000 + 1 entries. With delta only maintained we can think of diff of only one object if present.

Reply:
I meant keeping a copy of the running configuration for each candidate (in the current design each candidate starts as a copy of the running configuration). You need that for better conflict resolution when there are multiple users editing the configuration at the same time (please read about `configure private`, `update`, `commit force`, `configuration compare` and other new commands [here](https://github.com/opensourcerouting/frr/wiki/Transactional-CLI#config-modes)).

In any case, benchmarks have shown that our biggest bottleneck now isn't comparing configurations, but validating them. If we have 10,000 commands in the startup configuration, libyang needs to perform 10,000 full tree validations, which is *slow* (except when `--tcli` is used). Work is being done in libyang to cache certain validation checks and improve the performance of `lyd_validate()`. We can discuss more about this later!